File: src/GUI_Microservices/GUI_Crypto/ApiContracts/Requests/KlineData/KlineDataRequest.cs
using SharedLibrary.Enums;

namespace GUI_Crypto.ApiContracts.Requests.KlineData;

/// <summary>
/// Represents the parameters required to request Kline (candlestick) data from an exchange.
/// </summary>
public class KlineDataRequest
{
    /// <summary>
    /// The id of the trading pair.
    /// </summary>
    public required int IdTradingPair { get; set; }

    /// <summary>
    /// The main coin of the trading pair for which Kline data is requested.
    /// </summary>
    public required KlineDataRequestCoin CoinMain { get; set; }

    /// <summary>
    /// The quote coin of the trading pair.
    /// </summary>
    public required KlineDataRequestCoin CoinQuote { get; set; }

    /// <summary>
    /// The exchanges on which this trading pair is traded.
    /// </summary>
    public required IEnumerable<Exchange> Exchanges { get; set; }

    /// <summary>
    /// The interval for each Kline. All supported intervals can be found in the TimeFrame enum.
    /// </summary>
    public ExchangeKlineInterval Interval { get; set; }

    /// <summary>
    /// The optional start time for the Kline data.
    /// </summary>
    public DateTime StartTime { get; set; }

    /// <summary>
    /// The optional end time for the Kline data.
    /// </summary>
    public DateTime EndTime { get; set; }

    /// <summary>
    /// The maximum number of Kline records to retrieve for each trading pair. Maximum value is 1000 (default).
    /// </summary>
    public int Limit { get; set; } = 1000;
}

File: src/GUI_Microservices/GUI_Crypto/wwwroot/ts/chart/interfaces/kline-data-request.ts
export enum ExchangeKlineInterval {
    OneMinute = 1,
    FiveMinutes = 5,
    FifteenMinutes = 15,
    ThirtyMinutes = 30,
    OneHour = 60,
    FourHours = 240,
    OneDay = 1440,
    OneWeek = 10080,
    OneMonth = 43200,
}

export interface KlineDataRequestCoin {
    id: number;
    symbol: string;
    name: string;
}

export enum Exchange {
    Binance = 1,
    Bybit = 2,
    Mexc = 3,
}

export interface KlineDataRequest {
    idTradingPair: number;
    coinMain: KlineDataRequestCoin;
    coinQuote: KlineDataRequestCoin;
    exchanges: Exchange[];
    interval: ExchangeKlineInterval;
    startTime: string;
    endTime: string;
    limit: number;
}
